service/core/kafka/server/src/main/scala/com/lightbend/lagom/internal/broker/kafka/Producer.scala

```diff
@@ -85,57 +85,80 @@ private[lagom] object Producer {
     override def receive = {
       case EnsureActive(tag) =>
         offsetStore.prepare(s"topicProducer-$topicId", tag) pipeTo self
-        context.become(preparing(tag))
-    }
-
-    private def preparing(tag: String): Receive = {
-      case dao: OffsetDao =>
-        val readSideSource = eventStreamFactory(tag, dao.loadedOffset)
 
-        val (killSwitch, streamDone) = readSideSource
-          .viaMat(KillSwitches.single)(Keep.right)
-          .via(eventsPublisherFlow(dao))
-          .toMat(Sink.ignore)(Keep.both)
-          .run()
-
-        shutdown = Some(killSwitch)
-        streamDone pipeTo self
-        context.become(active)
+        kafkaConfig.serviceName match {
+          case None =>
+            context.become(waitingForOffsetDao(dao => run(tag, None, dao)))
+          case Some(name) =>
+            locateService(name) pipeTo self
+
+            // We locate the service and load the offset in parallel, so there's two possibilities, the offset could
+            // be loaded first, or the service could be located first. We combine the handlers for both using an
+            // orElse, and whichever one happens first causes us to switch to waiting for just the other one.
```

Comment:
The control flow seems a bit complicated. It's not doing anything other than waiting in the intermediate states, so could it zip the two futures together and then wait for the pair?

Something like

```scala
        val eventualOffsetDao =
          offsetStore.prepare(s"topicProducer-$topicId", tag)
        val eventualServiceUri =
          kafkaConfig.serviceName.fold(Future.successful(None: Option[URI]))(locateService(_))
        (eventualOffsetDao zip eventualServiceUri) pipeTo self

        context.become(initializing((dao, uri) => run(tag, uri, dao)))
```